crates/red_knot_python_semantic/src/types.rs

```diff
@@ -2423,6 +2464,56 @@ impl From<bool> for Truthiness {
     }
 }
 
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum Len {
+    /// Can be inferred statically.
+    Known(u64),
+    /// Cannot be inferred statically, falling back to `int`.
+    Unknown,
+    /// `len()` is known to raise an error.
+    Never,
```

Comment:
I understand the intention behind returning `Type::Never` in cases where we can statically see that the `len` call will return an error at runtime. But I'm not sure if this leads to the best user experience?

Once we have call signature checking, we will definitely raise a diagnostic for something like taking the `len` of a `Literal[1]`/`int` here:
```py
x = 1
xs = [x]

# some code

n = len(x)  # oops, was supposed to be `xs`

# more code
```

The question is what happens after the `n = len(x)` line. If we infer `Never`, we risk hiding further downstream errors. If we infer `int` (because we know that the error in the offending line has to be fixed anyway), we can catch other potential errors in the `more code` section.

Once we add unreachable-code analysis, we might even flag the `more code` section as unreachable if we infer `Never` here. Which is technically correct, but potentially confusing if all you need to do is to fix the `len(x)` error.